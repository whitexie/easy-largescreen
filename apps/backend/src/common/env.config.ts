/* eslint-disable node/prefer-global/process */
import * as dotenv from 'dotenv';

dotenv.config({ path: 'db.env' });

const env = process.env;

export default env;
