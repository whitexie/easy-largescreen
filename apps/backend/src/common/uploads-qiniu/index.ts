import * as qiniu from 'qiniu';
import env from '../env.config';

const { QINIU_ACCESS_KEY, QINIU_SECRET_KEY, QINIU_BUCKET, QINIU_EXPIRES } = env;
const mac = new qiniu.auth.digest.Mac(QINIU_ACCESS_KEY, QINIU_SECRET_KEY);

const options = {
  scope: QINIU_BUCKET,
  expires: Number(QINIU_EXPIRES) || 7200,
};

console.log('options => ', options);

const putPolicy = new qiniu.rs.PutPolicy(options);

export function getUploadToken() {
  const token = putPolicy.uploadToken(mac);
  const expires = Math.floor(Date.now() / 1000 + options.expires);
  return { token, expires };
}
