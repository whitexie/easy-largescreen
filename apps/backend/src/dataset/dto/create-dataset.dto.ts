import { IsNotEmpty, IsObject, IsString } from 'class-validator';

export class CreateDatasetDto {
  @IsString()
  @IsNotEmpty()
  name: string;

  @IsString()
  description: string;

  @IsObject()
  @IsNotEmpty()
  data: any;
}
