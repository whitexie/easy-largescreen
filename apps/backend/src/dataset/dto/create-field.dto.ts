import { ApiProperty, OmitType } from '@nestjs/swagger';
import { IsEnum, IsNotEmpty, IsString } from 'class-validator';
import { Field, ValueType } from '../entities/field.entity';

export class CreateFieldDto extends OmitType(Field, ['id', 'datasetId', 'created_time', 'updated_time', 'dataset'] as const) {
  @IsString()
  @IsNotEmpty({ message: 'fieldCode 不能为空' })
  @ApiProperty({ description: '字段code' })
  fieldCode: string;

  @IsEnum(ValueType, { message: 'valueType 不是一个有效值' })
  @ApiProperty({ description: '字段类型', enum: ValueType })
  valueType: ValueType;
}
