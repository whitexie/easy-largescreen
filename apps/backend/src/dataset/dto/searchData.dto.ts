import { Type } from 'class-transformer';
import { IsArray, IsEnum, IsNotEmpty, IsOptional, IsString, ValidateNested } from 'class-validator';

export enum CalculateType {
  SUM = 'sum',
  AVG = 'avg',
  COUNT = 'count',
}

class FieldDto {
  @IsNotEmpty()
  @IsString()
  fieldCode: string;

  @IsNotEmpty()
  @IsString()
  id: string;

  @IsOptional()
  @IsString()
  name?: string;

  @IsOptional()
  @IsString()
  valueType?: string;

  @IsOptional()
  @IsString()
  fieldType?: string;

  @IsOptional()
  @IsString()
  datasetId?: string;
}

class MetricFieldDto extends FieldDto {
  @IsNotEmpty()
  @IsEnum(CalculateType)
  calculateType: CalculateType;
}

export class SearchDataDto {
  @IsNotEmpty()
  @IsString()
  datasetId: string;

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => FieldDto)
  dimensionFields: FieldDto[];

  @IsArray()
  @ValidateNested({ each: true })
  @Type(() => MetricFieldDto)
  metricFields: MetricFieldDto[];
}
