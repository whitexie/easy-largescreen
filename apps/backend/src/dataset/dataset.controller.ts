import { Body, Controller, Post } from '@nestjs/common';
import { InjectEntityManager } from '@nestjs/typeorm';
import { EntityManager } from 'typeorm';
import { DatasetService } from './dataset.service';
import { CreateDatasetDto } from './dto/create-dataset.dto';
import { SearchDataDto } from './dto/searchData.dto';

@Controller('dataset')
export class DatasetController {
  constructor(
    private readonly datasetService: DatasetService,
    @InjectEntityManager() private entityManager: EntityManager,
  ) { }

  @Post()
  async create(@Body() createDatasetDto: CreateDatasetDto) {
    return this.datasetService.create(createDatasetDto);
  }

  @Post('/searchData')
  async searchData(@Body() params: SearchDataDto) {
    return this.datasetService.searchData(params);
  }
}
