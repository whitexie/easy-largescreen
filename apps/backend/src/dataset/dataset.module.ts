import { Module } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';
import { DatasetController } from './dataset.controller';
import { DatasetService } from './dataset.service';
import { Dataset } from './entities/dataset.entity';
import { Field } from './entities/field.entity';

@Module({
  imports: [TypeOrmModule.forFeature([Dataset, Field])],
  controllers: [DatasetController],
  providers: [DatasetService],
})
export class DatasetModule {}
