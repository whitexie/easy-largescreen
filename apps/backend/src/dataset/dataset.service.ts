import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { CreateDatasetDto } from './dto/create-dataset.dto';
import { SearchDataDto } from './dto/searchData.dto';
import { Dataset } from './entities/dataset.entity';

@Injectable()
export class DatasetService {
  constructor(
    @InjectRepository(Dataset)
    private datasetRepository: Repository<Dataset>,
  ) {}

  async create(createDatasetDto: CreateDatasetDto): Promise<Dataset> {
    const dataset = this.datasetRepository.create(createDatasetDto);
    return this.datasetRepository.save(dataset);
  }

  async searchData(searchParams: SearchDataDto) {
    const { datasetId, dimensionFields, metricFields } = searchParams;

    // 构建SQL查询
    const selectFields = dimensionFields.map(field =>
      `${field.fieldCode} AS "${field.id}"`,
    ).join(', ');

    const groupByFields = dimensionFields.map(field => field.fieldCode).join(', ');

    const aggregateFunctions = metricFields.map(field =>
      `${field.calculateType}(${field.fieldCode}) AS "${field.id}"`,
    ).join(', ');

    const sql = `
      SELECT ${selectFields}, ${aggregateFunctions}
      FROM ${datasetId}
      GROUP BY ${groupByFields}
    `;

    // console.log('SQL => ', sql);
    // 执行查询
    let results;
    try {
      results = await this.datasetRepository.manager.query(sql);
    }
    catch (error) {
      throw new HttpException({
        status: HttpStatus.INTERNAL_SERVER_ERROR,
        message: error.sqlMessage,
      }, HttpStatus.INTERNAL_SERVER_ERROR);
    }

    return results;
  }
}
