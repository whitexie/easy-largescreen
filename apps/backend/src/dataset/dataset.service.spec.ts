import { HttpException } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { getRepositoryToken } from '@nestjs/typeorm';
import { EntityManager, Repository } from 'typeorm';
import { DatasetService } from './dataset.service';
import { CalculateType, SearchDataDto } from './dto/searchData.dto';
import { Dataset } from './entities/dataset.entity';

describe('datasetService', () => {
  let service: DatasetService;
  let mockRepository: Partial<Repository<Dataset>>;

  beforeEach(async () => {
    mockRepository = {
      manager: {
        query: jest.fn(),
        connection: {} as any,
        transaction: jest.fn(),
        repositories: {},
        treeRepositories: {},
        plainObjectToEntityTransformer: {} as any,
        callAggregateFun: jest.fn(),
        createQueryBuilder: jest.fn(),
        hasId: jest.fn(),
        getId: jest.fn(),
        // 添加其他必要的EntityManager方法
      } as unknown as EntityManager,
    };

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        DatasetService,
        {
          provide: getRepositoryToken(Dataset),
          useValue: mockRepository,
        },
      ],
    }).compile();

    service = module.get<DatasetService>(DatasetService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('searchData', () => {
    it('应该成功执行查询并返回结果', async () => {
      const mockSearchParams: SearchDataDto = {
        datasetId: 'test_table',
        dimensionFields: [{ id: 'dim1', fieldCode: 'field1' }],
        metricFields: [{ id: 'metric1', fieldCode: 'field2', calculateType: CalculateType.SUM }],
      };
      const mockResults = [{ dim1: 'value1', metric1: 100 }];

      (mockRepository.manager.query as jest.Mock).mockResolvedValue(mockResults);

      const result = await service.searchData(mockSearchParams);

      expect(result).toEqual(mockResults);
      expect(mockRepository.manager.query).toMatchSnapshot();
    });

    it('应该在查询失败时抛出HttpException', async () => {
      const mockSearchParams: SearchDataDto = {
        datasetId: 'test_table',
        dimensionFields: [{ id: 'dim1', fieldCode: 'field1' }],
        metricFields: [{ id: 'metric1', fieldCode: 'field2', calculateType: CalculateType.SUM }],
      };

      (mockRepository.manager.query as jest.Mock).mockRejectedValue({ sqlMessage: 'Database error' });

      await expect(service.searchData(mockSearchParams)).rejects.toThrow(HttpException);
      await expect(service.searchData(mockSearchParams)).rejects.toThrow('Database error');
    });

    it('应该正确处理多个维度和指标字段', async () => {
      const mockSearchParams: SearchDataDto = {
        datasetId: 'test_table',
        dimensionFields: [
          { id: 'dim1', fieldCode: 'field1' },
          { id: 'dim2', fieldCode: 'field2' },
        ],
        metricFields: [
          { id: 'metric1', fieldCode: 'field3', calculateType: CalculateType.SUM },
          { id: 'metric2', fieldCode: 'field4', calculateType: CalculateType.AVG },
        ],
      };
      const mockResults = [{ dim1: 'value1', dim2: 'value2', metric1: 100, metric2: 50 }];

      (mockRepository.manager.query as jest.Mock).mockResolvedValue(mockResults);

      const result = await service.searchData(mockSearchParams);

      expect(result).toEqual(mockResults);
      expect(mockRepository.manager.query).toMatchSnapshot();
      expect(mockRepository.manager.query).toMatchSnapshot();
    });
  });
});
