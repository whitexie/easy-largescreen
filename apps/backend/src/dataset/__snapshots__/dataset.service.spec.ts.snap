// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datasetService searchData 应该成功执行查询并返回结果 1`] = `
[MockFunction] {
  "calls": [
    [
      "
      SELECT field1 AS "dim1", sum(field2) AS "metric1"
      FROM test_table
      GROUP BY field1
    ",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`datasetService searchData 应该正确处理多个维度和指标字段 1`] = `
[MockFunction] {
  "calls": [
    [
      "
      SELECT field1 AS "dim1", field2 AS "dim2", sum(field3) AS "metric1", avg(field4) AS "metric2"
      FROM test_table
      GROUP BY field1, field2
    ",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`datasetService searchData 应该正确处理多个维度和指标字段 2`] = `
[MockFunction] {
  "calls": [
    [
      "
      SELECT field1 AS "dim1", field2 AS "dim2", sum(field3) AS "metric1", avg(field4) AS "metric2"
      FROM test_table
      GROUP BY field1, field2
    ",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;
